### PR TITLE
Adds Rate class for #564

### PR DESCRIPTION
**Pull request by wayneparrott**
Implements a Rate timer based on the rclpy Rate timer. See issue #564.
Includes:
* update node, added node#createRate(hz)
* lib/rate.js
* types/rate.d.ts
* test/test-rate.js
* updated test/types/main.ts

Note for reviewers:
* See RateTimer in rate.js. This class runs a Timer in private rcl context.

### CODE REVIEW

**Review by minggangw — COMMENTED** [review]
Thanks for implementing Rate function, I didn't realize that we could instantiate another `ShadowNode` to create additional sub-thread, good idea. But I guess we still cannot block the Node.js main thread, see the comment. Shall we give a detailed explanation about what will happen when calling `sleep()`?

File: test/test-rate.js
Comment (minggangw) [inline test/test-rate.js:92]: I am wondering what it's the outcome if I insert:
```js
    setInterval(() => {
      // publish a top here.
    }, 10);
```

Comment (wayneparrott) [reply test/test-rate.js:92]: Here's a simple experiment to see the effect of rate.
Start publishing msgs every 1 sec from commandline using:
```
ros2 topic pub topic 'std_msgs/msg/String' '{data: hello}'
```
Run this example to received msgs every 2 seconds:
```javascript
'use strict';

const rclnodejs = require('./index.js');

async function main() {
  await rclnodejs.init();
  const node = rclnodejs.createNode('test_node');
  const subscriptions = node.createSubscription(
    'std_msgs/msg/String',
    'topic', 
    undefined,
    msg => console.log(`Received(${Date.now()}): ${msg.data}`)
  );
  const rate = node.createRate(0.5);
  let forever = true;
  while (forever) { // should run every 2 secs
    await rate.sleep();
    rclnodejs.spinOnce(node);
  }
}

main();
```

Comment (minggangw) [reply test/test-rate.js:92]: Yes, the subscription can work as expected, but publishing a topic can still be called through a callback of the `setInterval()`

Comment (wayneparrott) [reply test/test-rate.js:92]: >but publishing a topic can still be called through a callback of the setInterval()

Correct. Is that not what you were expecting? Are you expecting Rate.sleep() to govern, i.e., block, calls to publish()?

I coded up the example with the setInterval( ()=>publisher.publish('hello'),10). When the interval invokes the callback, the published msg is available to all subscribers who's node is spinning or spinOnce. In this example  the rate loop is  running at 1/2 hz, while a msg is published every 10ms by the setInterval. With that, every 200th msg will be received when spinOnce is called. I believe the 199 msgs received during rate.sleep() are lost.

File: lib/rate.js
Comment (mattrichard) [inline lib/rate.js]: Should the name be prefixed with the name of the node that created the Rate? Otherwise, it looks like if you ran 2 separate JS scripts that create a Rate the names would conflict since NODE_CNT wouldn't be shared.

File: test/test-rate.js
Comment (minggangw) [reply test/test-rate.js:92]: >>but publishing a topic can still be called through a callback of the setInterval()

>Correct. Is that not what you were expecting? Are you expecting Rate.sleep() to govern, i.e., block, calls to publish()?

The behavior (not block the loop) here is reasonable in Node.js, I just want to ensure when developers are calling `sleep()`, they are clear about the expectation (not a real sleep), thanks!

File: lib/rate.js
Comment (wayneparrott) [reply lib/rate.js]: Yes, absolutely! Good suggestion.
And/or we could make the Node hidden altogether.

**Comment by wayneparrott** [comment]
PROPOSAL - refactor the Rate design as follows:
1. rename RateTimer to RateTimerServer, to produce Timers for use by Rate instances 
2. move RateTimerServer management from each rate instance into Node
3. use lazy init RateTimerServer on first Node#createRate() call
4. revamp Rate back to my 1st version where it is initialized with a Timer from the RateTimerServer during Node#createRate()

Tradeoffs of this refactor include:
1. Only 1 rcl context and node to support for all Rates created by a node - fewer moving parts, increases efficiency
2. Any RateTimerServer node initialization, etc is hidden from developers working with rate - simplifies use 
3. Increases Node's responsibility and internal complexity

Apologies for pushing this proposal after you've invested time reviewing the current impl. If approved I prefer making this change immediately before merging. Doing it in the future will result in a breaking api change for Rate.

Thoughts?

**Comment by minggangw** [comment]
The redesign sounds more reasonable, I like using only one node to do it. We could also land this PR in advance and create a new issue to target the refactoring, as the changes are on `develop` I think it's fine, it's up to you. Please go ahead, thanks a lot!

**Comment by wayneparrott** [comment]
The latest commit #b0fedaa is the refactored version of Rate with Node now managing a lazy initialized RateTimerServer, 1 per node instance. The RateTimerServer class is located in node.js. 

Per @mattrichard suggestion the RateTimerServer manages a private rcl context and node. This node shares the name of its container node as `_$(containing_node.name())_rate_timer_server`. The '_' prefix will keep it hidden from commandline operations such as `ros2 node list`.

Apologies for the late refactor after you've already invested review time. I'm much more positive about this version that the former.

**Review by minggangw — APPROVED** [review]
Thanks for refactoring the Rate implementation quickly, LGTM @mattrichard any comments? If everything is OK, I'm going to merge this PR, thanks!

File: lib/node.js
Comment (mattrichard) [inline lib/node.js]: Since this only relates to Rates, can this class go in the `rate.js` file instead? The `node.js` file is already very large.

Other than that it LGTM.

Comment (wayneparrott) [reply lib/node.js]: Agree node.js is getting large and I originally I implemented RateTimeServer in rate.js for that reason. But at the last min I moved it to node.js to keep it private and avoid exporting it from Rate. I'm fine with moving RateTimeServer to rate and exporting it.

Comment (wayneparrott) [reply lib/node.js]: The latest commit includes RateTimeServer relocated from node.js to rate.js

**Review by mattrichard — APPROVED** [review]
LGTM

**Pull request merged**
